File: backend/app/api/v1/analyses.py
"""Analyses REST API."""

from __future__ import annotations

from datetime import date

from fastapi import APIRouter, Depends, HTTPException, Query, status

from app.dependencies import (
    get_analysis_norms_service,
    get_analysis_service,
    get_current_user,
)
from app.models.user import User
from app.schemas.analysis import (
    AnalysisHistoryResponse,
    AnalysisParameterDetail,
    AnalysisParameterNorm,
    AnalysisParameterSummary,
    AnalysisRecordCreate,
    AnalysisRecordFull,
    AnalysisRecordSummary,
    AnalysisRecordUpdate,
    AnalysisValueResponse,
    AnalysisValueUpdate,
)
from app.services.analysis_norms_service import AnalysisNormsService
from app.services.analysis_service import (
    AnalysisRecordNotFoundError,
    AnalysisService,
    AnalysisValueNotFoundError,
)

router = APIRouter(prefix="/v1/analyses", tags=["analyses"])


# ---------- Records -------------------------------------------------------- #


@router.get("", response_model=list[AnalysisRecordSummary])
async def list_analyses(
    date_from: date | None = Query(default=None),
    date_to: date | None = Query(default=None),
    only_abnormal: bool = Query(default=False),
    current_user: User = Depends(get_current_user),
    svc: AnalysisService = Depends(get_analysis_service),
) -> list[AnalysisRecordSummary]:
    return await svc.list_records(
        current_user.id,
        date_from=date_from,
        date_to=date_to,
        only_abnormal=only_abnormal,
    )


@router.post(
    "", response_model=AnalysisRecordFull, status_code=status.HTTP_201_CREATED
)
async def create_analysis(
    data: AnalysisRecordCreate,
    current_user: User = Depends(get_current_user),
    svc: AnalysisService = Depends(get_analysis_service),
) -> AnalysisRecordFull:
    return await svc.create_record(current_user.id, data)


@router.get("/{record_id}", response_model=AnalysisRecordFull)
async def get_analysis(
    record_id: int,
    current_user: User = Depends(get_current_user),
    svc: AnalysisService = Depends(get_analysis_service),
) -> AnalysisRecordFull:
    try:
        return await svc.get_record(current_user.id, record_id)
    except AnalysisRecordNotFoundError as exc:
        raise HTTPException(status_code=404, detail="analysis_not_found") from exc


@router.patch("/{record_id}", response_model=AnalysisRecordFull)
async def update_analysis(
    record_id: int,
    data: AnalysisRecordUpdate,
    current_user: User = Depends(get_current_user),
    svc: AnalysisService = Depends(get_analysis_service),
) -> AnalysisRecordFull:
    try:
        return await svc.update_record(current_user.id, record_id, data)
    except AnalysisRecordNotFoundError as exc:
        raise HTTPException(status_code=404, detail="analysis_not_found") from exc


@router.delete("/{record_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_analysis(
    record_id: int,
    current_user: User = Depends(get_current_user),
    svc: AnalysisService = Depends(get_analysis_service),
) -> None:
    try:
        await svc.delete_record(current_user.id, record_id)
    except AnalysisRecordNotFoundError as exc:
        raise HTTPException(status_code=404, detail="analysis_not_found") from exc


# ---------- Values --------------------------------------------------------- #


@router.patch("/values/{value_id}", response_model=AnalysisValueResponse)
async def update_value(
    value_id: int,
    data: AnalysisValueUpdate,
    current_user: User = Depends(get_current_user),
    svc: AnalysisService = Depends(get_analysis_service),
) -> AnalysisValueResponse:
    try:
        return await svc.update_value(current_user.id, value_id, data)
    except AnalysisValueNotFoundError as exc:
        raise HTTPException(status_code=404, detail="value_not_found") from exc


# ---------- Parameter history --------------------------------------------- #


@router.get(
    "/parameters/{parameter_code}/history",
    response_model=AnalysisHistoryResponse,
)
async def parameter_history(
    parameter_code: str,
    date_from: date | None = Query(default=None),
    date_to: date | None = Query(default=None),
    current_user: User = Depends(get_current_user),
    svc: AnalysisService = Depends(get_analysis_service),
) -> AnalysisHistoryResponse:
    return await svc.get_parameter_history(
        current_user.id,
        parameter_code,
        date_from=date_from,
        date_to=date_to,
    )


# ---------- Dictionary --------------------------------------------------- #


dict_router = APIRouter(prefix="/v1/dictionaries", tags=["dictionaries"])


@dict_router.get(
    "/analysis-parameters",
    response_model=list[AnalysisParameterSummary],
)
async def search_parameters(
    search: str = Query(default="", max_length=100),
    limit: int = Query(default=20, ge=1, le=50),
    _current_user: User = Depends(get_current_user),
    norms: AnalysisNormsService = Depends(get_analysis_norms_service),
) -> list[AnalysisParameterSummary]:
    summaries = await norms.search(search, limit=limit)
    return [
        AnalysisParameterSummary(
            code=s.code, name_ru=s.name_ru, unit=s.unit, category=s.category
        )
        for s in summaries
    ]


@dict_router.get(
    "/analysis-parameters/{code}",
    response_model=AnalysisParameterDetail,
)
async def get_parameter_detail(
    code: str,
    gender: str = Query(default="any"),
    age: int = Query(default=18, ge=0, le=150),
    _current_user: User = Depends(get_current_user),
    norms: AnalysisNormsService = Depends(get_analysis_norms_service),
) -> AnalysisParameterDetail:
    param = norms.get(code)
    if param is None:
        raise HTTPException(status_code=404, detail="parameter_not_found")
    norm = norms.get_norm(code, gender=gender, age=age)
    return AnalysisParameterDetail(
        code=param.code,
        name_ru=param.name_ru,
        name_en=param.name_en,
        unit=param.unit,
        alternative_units=param.alternative_units,
        category=param.category,
        description=param.description,
        synonyms=param.synonyms,
        applicable_norm=(
            AnalysisParameterNorm(
                min=norm.min,
                max=norm.max,
                gender=norm.gender,
                age_from=norm.age_from,
                age_to=norm.age_to,
            )
            if norm is not None
            else None
        ),
    )

File: frontend/src/api/analyses.ts
import apiClient from './client'

export type AbnormalType = 'low' | 'high' | 'normal' | 'unknown'

export interface AnalysisValueInput {
  parameter_code: string | null
  parameter_name: string
  value: string
  unit: string
  reference_min: string | null
  reference_max: string | null
}

export interface AnalysisValue {
  id: number
  record_id: number
  user_id: number
  parameter_code: string | null
  parameter_name: string
  value: string
  unit: string
  reference_min: string | null
  reference_max: string | null
  is_abnormal: boolean
  abnormal_type: AbnormalType
}

export interface AnalysisRecordSummary {
  id: number
  user_id: number
  analysis_date: string
  lab_name: string | null
  values_count: number
  abnormal_count: number
}

export interface AnalysisRecordFull {
  id: number
  user_id: number
  analysis_date: string
  lab_name: string | null
  doctor_referral: string | null
  notes: string | null
  created_at: string
  values: AnalysisValue[]
}

export interface AnalysisRecordCreate {
  analysis_date: string
  lab_name: string | null
  doctor_referral: string | null
  notes: string | null
  values: AnalysisValueInput[]
}

export interface AnalysisHistoryPoint {
  analysis_date: string
  value: string
  value_numeric: number | null
  unit: string
  reference_min: string | null
  reference_max: string | null
  is_abnormal: boolean
  abnormal_type: AbnormalType
}

export interface AnalysisHistory {
  parameter_code: string
  parameter_name: string | null
  unit: string | null
  points: AnalysisHistoryPoint[]
}

export interface AnalysisParameterSummary {
  code: string
  name_ru: string
  unit: string
  category: string
}

export interface AnalysisParameterDetail extends AnalysisParameterSummary {
  name_en: string
  alternative_units: string[]
  description: string | null
  synonyms: string[]
  applicable_norm: {
    min: number | null
    max: number | null
    gender: string
    age_from: number
    age_to: number
  } | null
}

export interface AnalysesFilters {
  date_from?: string
  date_to?: string
  only_abnormal?: boolean
}

export const analysesApi = {
  list(filters: AnalysesFilters = {}) {
    return apiClient.get<AnalysisRecordSummary[]>('/v1/analyses', { params: filters })
  },
  get(id: number) {
    return apiClient.get<AnalysisRecordFull>(`/v1/analyses/${id}`)
  },
  create(payload: AnalysisRecordCreate) {
    return apiClient.post<AnalysisRecordFull>('/v1/analyses', payload)
  },
  update(id: number, payload: Partial<Pick<AnalysisRecordCreate, 'lab_name' | 'doctor_referral' | 'notes' | 'analysis_date'>>) {
    return apiClient.patch<AnalysisRecordFull>(`/v1/analyses/${id}`, payload)
  },
  remove(id: number) {
    return apiClient.delete<void>(`/v1/analyses/${id}`)
  },
  updateValue(valueId: number, payload: Partial<AnalysisValueInput>) {
    return apiClient.patch<AnalysisValue>(`/v1/analyses/values/${valueId}`, payload)
  },
  history(parameterCode: string, dateFrom?: string, dateTo?: string) {
    return apiClient.get<AnalysisHistory>(
      `/v1/analyses/parameters/${parameterCode}/history`,
      { params: { date_from: dateFrom, date_to: dateTo } },
    )
  },
  searchParameters(query: string) {
    return apiClient.get<AnalysisParameterSummary[]>(
      '/v1/dictionaries/analysis-parameters',
      { params: { search: query, limit: 20 } },
    )
  },
  parameterDetail(code: string, gender: string, age: number) {
    return apiClient.get<AnalysisParameterDetail>(
      `/v1/dictionaries/analysis-parameters/${code}`,
      { params: { gender, age } },
    )
  },
}
